### ISSUE TITLE
Does jsx-no-bind contradicts with no-string-refs?

### ISSUE THREAD

**Issue by guocheng** [yannickcr/eslint-plugin-react#524]
The `no-string-refs` document provides the following example:

```js
render() {
    return <div ref={c => this.hello = c}>Hello, world.</div>;
  }
```

I am curious if the arrow function `ref={c => this.hello = c}` gets recreated every time render is called and thus contradicts with the `jsx-no-bind` rule.

Thanks.

**Comment by anthonator** [comment]
I'm curious about this too!

**Comment by ljharb** [comment]
Yes, it would, unless you've enabled the override options to it for arrow functions and/or refs.

**Comment by Kovensky** [comment]
How does React actually behave if the identity of the `ref` pseudo-prop changes?

If you put it on a DOM element, it's not a prop that is reflected on the DOM, so even if the diff algorithm finds that it needs to do "something" because it changed, it would not ultimately write to the DOM.

If you put it on a custom component, it's not a prop that can be read by the component (AFAIK), so it shouldn't do the full `shouldComponentUpdate` / `render` cycle (if it does, even though it's a prop that the component cannot do anything with, it sounds like a React bug).

**Comment by cristian-sima** [comment]
+1

**Comment by ljharb** [comment]
@cristian-sima please don't reply "+1" to posts on github - that's what "emoji reactions on the original post" are for. I'll continue to clean up any such posts and I encourage you to add a +1 to the original post.

**Comment by ixisio** [comment]
@Kovensky, @ljharb, @anthonator, @guocheng 

[from eslint docs](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md): 
> A bind call or arrow function in a JSX prop will create a brand new function on every single render. This is bad for performance, as it will result in the garbage collector being invoked way more than is necessary.

So, try it like that:

````js
export class Test extends React.Component {
  constructor(props) {
    super(props);

    this.elementRef = this.elementRef.bind(this);
  }

  elementRef (element) {
    this.element = element;
  }
  
  render () {
    return (
        <div ref={this.elementRef}>Hello</div>
    )
  }
}
```

Cheers Andy

**Comment by anthonator** [comment]
We've been doing something similar but slightly different:

```js
export class Test extends React.Component {
  constructor(props) {
    super(props);
  }

  elementRef (element) {
    this.element = element;
  }

  elementRef = this.elementRef.bind(this)

  render () {
    return (
        <div ref={this.elementRef}>Hello</div>
    )
  }
}
```

**Comment by ljharb** [comment]
@anthonator that's exactly the best solution (you can delete your useless constructor tho)

**Comment by ixisio** [comment]
@guocheng So, solution is clear. Please close this issue... thanks
